### ISSUE TITLE
False positive CA1508, condition is always true

### ISSUE THREAD

**Issue by myblindy** [dotnet/roslyn-analyzers#5254]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `'encodedBytes is null' is always 'true'. Remove or refactor the condition(s) to avoid dead code.`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 6 SDK](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: SDK 6.0.100-preview.6.21355.2

### Describe the bug

The analyzer seems confused by async tasks:

```C#
using System;
using System.Threading.Tasks;

byte[] encodedBytes = default;
await Task.Run(() => encodedBytes = new byte[1]).ConfigureAwait(false);
if (encodedBytes is null)
    Console.WriteLine();
```

### Expected behavior

I would expect not to see that warning in this case, either because the analyzer can track the variable assignment (this is a relatively simple pattern of a fully awaited task that definitely assigns to the variable), or because it's too complex and can't tell for sure. 

### Actual behavior

Instead, the analyzer seems confident that the assignment never takes place.

**Comment by RikkiGibson** [comment]
It feels like this is also a limitation of the built-in flow analysis. The trouble is we don't know "when" a lambda was invoked. [SharpLab](https://sharplab.io/#v2:CYLg1APgxAdgrgGwQQwEYIKYAIMzZgWAChiABABi1IEYBWAbmLIGYqAmLAYSwG9isBVVqQAsWALIAKAJS9+ghTXIB+LAA8sAXiwAiZDsZEFigGwA6AEpwYkmVoB86rVnhJph44LVmAKgHsAZQAXACcASxgAcxkPQQBfYgSiIA===)

```cs
#nullable enable

using System.Threading.Tasks;

public class C {
    public void M() {
        string? x = "a";
        Task.Run(() => x = null);
        x.ToString(); // no warning is given, but perhaps one should be given
    }
}
```

We could pessimistically assume that it might have been invoked at the point it was used. But we might also want to be careful with the subtleties of that. [SharpLab](https://sharplab.io/#v2:CYLg1APgxAdgrgGwQQwEYIKYAIMzZgWAChiABAJgEZiyBmLCrAYSwG9itOH7SAWLALIAKAJRsOXSaUoAGAPxYAHlgC8WAETJ1AbgmTOAmUNGqAfEtVZ4SADRKAdABUA9gGUALgCcAljADmoiK6RJIAvjQhXKQ8/IZC0uRYyADG7t7OMHbSMlgAnmKsSanpMKLaWOFEoUA===)
```cs
#nullable enable

using System;

public class C {
    public void M() {
        string? x = "a";
        M0(
            () => x = null,
            x.ToString()); // we probably don't want to warn here

        x.ToString(); // but we would warn here
    }

    public void M0(Action action, string y) { action(); }
}
```

There are still holes if a lambda escaped from the method we passed it to and it was invoked at some arbitrary later time. But it does seem like the "pessimism" of assuming a lambda might have been invoked at the point where it is used could be useful. [SharpLab](https://sharplab.io/#v2:CYLg1APgxAdgrgGwQQwEYIKYAIMzZgWAChiABAJgEZiyBmLCrAYSwG9itOsAHAJwEsAbsgAu2UpQBsDSuSwB9ZAGMR/APYwA3By6l6E6aQAsWAMoYRE8gAorWZao0BKLAF4AfAofqYb+yp9tIi4sHU4KOXZgkN1KAAYAfiwADz8AImQ0oJiuc0tZa2sXDxS/eCQnbJzS1yw01CywmMUAjSKqnOSAOgAVNVMRARgAc3asAHpxrAB3bGmNAHIRGeReXwALDF5sVDhlkXX+AGcZ/iQsADkAJQBRJoBfYkeiIA==)

```cs
#nullable enable

using System;

public class C {
    private static Action _action;
    public static void SetAction(Action action) => _action = action;
    
    public static void Main() {
        string? x = "a";
        SetAction(() => x = null);
        x = "b";
        _action();
        x.ToString(); // we won't warn here but this will NRE
    }
}
```

@cston do you think it's worthwhile to make any revision to nullable and region analysis to make it so state changes at the end of a lambda may be "joined" to the state of the containing function?

**Comment by myblindy** [comment]
> We could pessimistically assume that it might have been invoked at the point it was used

I feel like that is much preferable to a warning who's (manual) fix is to delete the code, thus breaking your code. If you don't know, you don't know, everyone knows [computing science has hard limits on static analysis](https://en.wikipedia.org/wiki/Halting_problem), but don't assume the code is broken and highlight it for deletion. 

> There are still holes if a lambda escaped from the method we passed it to and it was invoked at some arbitrary later time

That said, I would point out that the code in question is not a generic piece of code that calls a function and passes in a lambda function, it's a task being awaited, one of the pillars of modern C#. If there is one heuristic that could be added, I would argue that that would be the one.

**Comment by myblindy** [comment]
Also your first code example doesn't await the task, so it won't crash with an NRE. Perhaps a different warning would be better there, but definitely not CA1508.

**Comment by jmarolf** [comment]
> @cston do you think it's worthwhile to make any revision to nullable and region analysis to make it so state changes at the end of a lambda may be "joined" to the state of the containing function?

imho this falls into the "threading rules" bucket and I am ok not changing the analysis rules here. The unfortunate thing is that the analyzer for CA1508 probably needs to take a different tactic than the compiler does is flow analysis due to the UX of how it reports things.

If a user writes this code and it null-refs then I think its reasonable to say "Nullable analysis doesn't handle value assignment across threads". However it is annoying to be told some code isn't used if it is used in other threads.  

Essentially: people notice false positives but not false negatives. The C# compiler not warning about possible null assignment _could_ be a false negative (I'll leave that to @RikkiGibson and @cston to decide) but I don't think anyone is going to complain about it. In this case we should bail on dead code analysis with lambdas and extension methods present for cases where we know control flow affects the output.

**Comment by RikkiGibson** [comment]
It looks like the analyzer actually works by reading a control flow graph. I would have to look more closely to see where the fix actually belongs (in the compiler or over here). But it might still be interesting to see if an analogous fix is applicable for region analysis and nullable.

**Comment by jmarolf** [comment]
@RikkiGibson I believe it is using the dataflow APIs which are from [here](https://github.com/dotnet/roslyn-analyzers/tree/8ca2a798f047e45600e98ed8e52a1b686f7c3ab8/src/Utilities/FlowAnalysis/FlowAnalysis) and work differently from the compiler

**Comment by mavasani** [comment]
This is just a bug somewhere in the DFA framework in this repo. The analysis needs to pessimistically bail out for this case.

**Comment by myblindy** [comment]
> The analysis needs to pessimistically bail out for this case.

Exactly, that's what I've been saying as well!

**Comment by crfrolik** [comment]
There are many issues like this (false positive bugs) that have been open for years.
I don't mean to be rude, but these false positives cause issues for lots of developers. Is there any plan to fix them?

**Comment by mavasani** [comment]
The original false positive in the issue description has been fixed with #6811. We can certainly make the analysis more aggressive and smarter to avoid more false negatives, but that needs more investment, performance measurements to ensure we don't regress on performance to gain more precision. Please file separate issue(s) for further flow analysis enhancements for CA1508.
